plasmapy/diagnostics/thomson.py

```diff
@@ -141,177 +357,535 @@ def spectral_density(
+
+    # **************
+    # Electron velocity
+    # **************
+    electron_speed = np.zeros(num_e)
+    for num in range(num_e):
+        k = f"electron_speed_{num}"
+        if k in params:
+            electron_speed[num] = params[k].value
+        else:
+            # electron_speed[e] = 0 already
+            params.add(k, value=0, vary=False)
+
+    if "electron_vdir" not in settings:
+        if np.all(electron_speed == 0):
+            # vdir is arbitrary in this case because vel is zero
+            settings["electron_vdir"] = np.ones([num_e, 3])
+        else:
+            raise ValueError(
+                "Key 'electron_vdir' must be defined in kwarg 'settings' if "
+                "any electron population has a non-zero speed (i.e. any "
+                "params['electron_speed_<#>'] is non-zero)."
+            )
+    norm = np.linalg.norm(settings["electron_vdir"], axis=-1)
+    settings["electron_vdir"] = settings["electron_vdir"] / norm[:, np.newaxis]
+
+    # **************
+    # Ion velocity
+    # **************
+    ion_speed = np.zeros(num_i)
+    for num in range(num_i):
+        k = f"ion_speed_{num}"
+        if k in params:
+            ion_speed[num] = params[k].value
+        else:
+            # ion_speed[i] = 0 already
+            params.add(k, value=0, vary=False)
+
+    if "ion_vdir" not in list(settings.keys()):
+        if np.all(ion_speed == 0):
+            # vdir is arbitrary in this case because vel is zero
+            settings["ion_vdir"] = np.ones([num_i, 3])
+        else:
+            raise ValueError(
+                "Key 'ion_vdir' must be defined in kwarg 'settings' if "
+                "any ion population has a non-zero speed (i.e. any "
+                "params['ion_speed_<#>'] is non-zero)."
+            )
+    norm = np.linalg.norm(settings["ion_vdir"], axis=-1)
+    settings["ion_vdir"] = settings["ion_vdir"] / norm[:, np.newaxis]
+
+    if "instr_func" not in settings or settings["instr_func"] is None:
+        settings["instr_func_arr"] = None
+    else:
+        # Create instr_func array from instr_func
+        instr_func = settings["instr_func"]
+        wspan = (np.max(wavelengths) - np.min(wavelengths)) / 2
+        eval_w = np.linspace(-wspan, wspan, num=wavelengths.size)
+        instr_func_arr = instr_func(eval_w * u.m)
+
+        if type(instr_func_arr) != np.ndarray:
+            raise ValueError(
+                "instr_func must be a function that returns a "
+                "np.ndarray, but the provided function returns "
+                f" a {type(instr_func_arr)}"
+            )
+
+        if wavelengths.shape != instr_func_arr.shape:
+            raise ValueError(
+                "The shape of the array returned from the "
+                f"instr_func ({instr_func_arr.shape}) "
+                "does not match the shape of the wavelengths "
+                f"array ({wavelengths.shape})."
+            )
+
+        instr_func_arr *= 1 / np.sum(instr_func_arr)
```

Comment:
I propose that we pronounce the `*=` operator as "starquals".  😬

Reply:
+= "plusquals"
-= "minquals"
But what about /=...

Reply:
I was completely seriously thinking "addquals", "subquals", "multiquals", and "divquals" completely seriously without any seriouslessness at all.